### DISCUSSION TITLE
v0.6.0

### DISCUSSION THREAD

**Discussion by takatost** [langgenius/dify#3163] [Releases]
## Introducing Dify Workflow! 🎉
The much-anticipated workflow feature is here: In a nutshell, workflow provides a visual canvas for defining complex tasks as smaller, manageable steps (nodes). This reduces reliance on prompt engineering and LLM agent capabilities, taking the stability and reproducibility of your LLM applications to the next level by letting you be in control.
**There are two Workflow application types with this update:**
- **Workflow App**
  Targeting Automation and Batch Processing: This is ideal for translation, data analysis, content generation, email automation, and more.
  ![image](https://github.com/langgenius/dify/assets/5485478/67c214e1-bc7e-44ad-8d81-297a30ebcc99)

- **Chatflow App (A Sub-Type of Chatbot)**
  For Conversational Applications: Suitable for customer service, semantic search, and more conversational apps requiring multi-step logic in crafting the response. 
  Compared to the regular Workflow app type, Chatflow adds chat-specific features such as conversation history support (Memory), tagged replies, an Answer node type for streaming responses, and support for rich text and images.
  
  ![image](https://github.com/langgenius/dify/assets/5485478/3fe94294-7e44-497a-9f19-e89808b2675f)

For more information, please visit: https://docs.dify.ai/features/workflow/introduce

Other Enhancements:
- Optimized UI flow for app creation.
- Conversion support from various basic application types to Workflow-based applications.
  > - Basic / Expert mode Chatbot apps → Chatflow
  > - Text Generator → Workflow
  >
  >   ![image](https://github.com/langgenius/dify/assets/5485478/9227958e-4a9d-482b-91b8-7c5b72a16a14)

- Dify's official app templates are now available in self-hosted mode.
- Support for adding descriptions to applications.
- Support for porting applications in and out of Dify with DSL.
- Under the hood, we also refactored the underlying execution logic of all app types for cleaner architecture and a tidier repo.

## Update Guide

**If you need to upgrade from** `0.6.0-preview-workflow.1`**, you will need to connect to** **PostgreSQL** **and execute the following SQL (migration inserted in the main branch) to ensure data integrity.**

```SQL
ALTER TABLE dataset_keyword_tables ADD COLUMN data_source_type VARCHAR(255) NOT NULL DEFAULT 'database';
ALTER TABLE embeddings ADD COLUMN provider_name VARCHAR(40) NOT NULL DEFAULT '';
ALTER TABLE embeddings DROP CONSTRAINT embedding_hash_idx;
ALTER TABLE embeddings ADD CONSTRAINT embedding_hash_idx UNIQUE (model_name, hash, provider_name);
```

### Docker compose deployments:

1. Get the latest code from the main branch:

   ```Shell
   git checkout main
   git pull origin main
   ```

2. Go to the next step and update to the latest image:

   ```Shell
   cd docker
   docker compose up -d
   ```

3. We also moved the agent data within the database, **Execute** **the below script** **to complete the migrate:** **(NEW)**

   ```Shell
   docker compose exec api flask convert-to-agent-apps
   ```

### Source Code deployments:

1. Stop API server, Worker and Web frontend Server.

2. Get the latest code from the main branch:

   ```Shell
   git checkout main
   git pull origin main
   ```

3. Update Python dependencies:

   ```Shell
   cd api
   pip install -r requirements.txt
   ```

4. Then, let's run the migration script:

   ```Shell
   flask db upgrade
   ```

5. We also moved the agent data within the database, **Execute the below script to complete the migrate: (NEW)**

   ```Shell
   flask convert-to-agent-apps
   ```

6. Finally, run API server, Worker and Web frontend Server again.

## What's Changed
* Update README.md by @guchenhe in https://github.com/langgenius/dify/pull/3081
* add segment function billing check for SAAS env by @JohnJyong in https://github.com/langgenius/dify/pull/3082
* feat (new llm): add support for openrouter by @Korayem in https://github.com/langgenius/dify/pull/3042
* fix(duckduckgo-search): invoke error by @QIN2DIM in https://github.com/langgenius/dify/pull/3077
* Chore/move chrome ext by @crazywoola in https://github.com/langgenius/dify/pull/3085
* add embedding cache and clean embedding cache job by @JohnJyong in https://github.com/langgenius/dify/pull/3087
* new readme slogan by @guchenhe in https://github.com/langgenius/dify/pull/3094
* fix: update show names for supported file types of xlsx and docx by @bowenliang123 in https://github.com/langgenius/dify/pull/3091
* fix typo by @leslie2046 in https://github.com/langgenius/dify/pull/3098
* fix typo in readme by @jat001 in https://github.com/langgenius/dify/pull/3096
* fix unstructured api,remove unused parameters by @vikeychen in https://github.com/langgenius/dify/pull/3056
* Improve ModelTypeEnum type by @lng2020 in https://github.com/langgenius/dify/pull/3051
* feat: claude3 tool call by @Yeuoly in https://github.com/langgenius/dify/pull/3111
* chore: replace outdated config in vscode debug settings by @lng2020 in https://github.com/langgenius/dify/pull/3106
* feat: update link by @crazywoola in https://github.com/langgenius/dify/pull/3121
* chore: update link of feedback by @JzoNgKVO in https://github.com/langgenius/dify/pull/3130
* Duplicate embedding cache check by @JohnJyong in https://github.com/langgenius/dify/pull/3134
* improve qa generate prompt by @JohnJyong in https://github.com/langgenius/dify/pull/3132
* fix: Turn off SWR automatic revalidation when window is focused by @Mxtzz in https://github.com/langgenius/dify/pull/3129
* Feat/update issue template by @crazywoola in https://github.com/langgenius/dify/pull/3147
* fix: chat app sometimes may crash by @iamjoel in https://github.com/langgenius/dify/pull/3151
* Update descriptions in StackExchange Tool by @richards199999 in https://github.com/langgenius/dify/pull/3043
* Fix: stop indexing status check when api of status checking failed by @JzoNgKVO in https://github.com/langgenius/dify/pull/3156
* Optimize csv and excel extract by @JohnJyong in https://github.com/langgenius/dify/pull/3155
* fix: knowledge create display error by @zxhlyh in https://github.com/langgenius/dify/pull/3157
* FEAT: NEW WORKFLOW ENGINE by @takatost in https://github.com/langgenius/dify/pull/3160

## New Contributors
* @Korayem made their first contribution in https://github.com/langgenius/dify/pull/3042
* @QIN2DIM made their first contribution in https://github.com/langgenius/dify/pull/3077
* @jat001 made their first contribution in https://github.com/langgenius/dify/pull/3096
* @Mxtzz made their first contribution in https://github.com/langgenius/dify/pull/3129

**Full Changelog**: https://github.com/langgenius/dify/compare/0.5.11-fix1...0.6.0

<hr /><em>This discussion was created from the release <a href='https://github.com/langgenius/dify/releases/tag/0.6.0'>v0.6.0</a>.</em>

**Comment by bowenliang123** [comment]
Congratulations! Dify brings us a new era of LLM agents.

**Reply by findigital** [reply]
This is awseome!

**Comment by JokerQyou** [comment]
Can a node (specifically the LLM node) have multiple entry points? I'm thinking of using chatflow to mimic function calling. It works like this:

- User input some question, it may involves querying some realtime data about the user
- an LLM node is provided with some function definitions, and this node decide that it needs to call `get_username` function, it outputs `FUNC:get_username()`
- an IF-ELSE node matches the `FUNC:` prefix, and pass it to a template node to trim the prefix, which outputs `get_username()`
- an HTTP node is invoked with function name as body, and external API returns the name of current user
- then I would like to pass the username back to previous LLM node, so it receives the return value of previous function call, and can decide to perform actions using the data.

This is basically a home-brewed agent mode chatbot (because currently dify's agent mode cannot be customized - the prompt is not editable, and the agent always replaces dataset context rather than append to it).

**Reply by takatost** [reply]
You might need a `reasoning` node, we will provide it in the next iteration.
